### DISCUSSION TITLE
Arbitrary simple (numeric) type conversion

### DISCUSSION THREAD

**Discussion by jacek-s** [nlohmann/json#3793] [Q&A]
I'm trying to get arbitrary type converstion to work with simple custom types that are not stucts or classes, based on documentation available here: https://json.nlohmann.me/features/arbitrary_types/

What I'm trying to do is as follows:
- I have a JSON field that can either have two "special" string values, let's say "string 1" and "string 2"
- If this field isn't set to one of those string values, it's otherwise always set to positive integer value
- In the code, I want to store this value in a single variable that either holds one of two special values that correspond to two possible string values or otherwise holds a positive integer value. I don't want to make it a struct/class.

My attempt:
```
#include <nlohmann/json.hpp>
#include <iostream>

using CustomType = int;

// This does not work at all (from_json not called)
//void to_json(nlohmann::json& j, const CustomType& val)
//{
//}
//void from_json(const nlohmann::json& j, CustomType& val)
//{
//  std::cout << "from_json called: " << j << std::endl;
//}

NLOHMANN_JSON_NAMESPACE_BEGIN
template<>
struct adl_serializer<CustomType>
{
  static void to_json(json& j, const CustomType& val)
  {
  }
  // This gets called for 'intVal' also
  static void from_json(const json& j, CustomType& opt)
  {
    std::cout << "from_json called: " << j << std::endl;
  }
};
NLOHMANN_JSON_NAMESPACE_END

int main(int argc, char* argv[])
{
  nlohmann::json json = nlohmann::json::parse(R"({"intVal": 123, "customVal": 456})");
  int intVal = json["intVal"];
  CustomType customVal = json["customVal"];
}
```

Output of the program above is:
> from_json called: 123
> from_json called: 456

My problems:
- The first version of `from_json` and `to_json` doesn't get called at all 
- The second version (adl_serializer specialization) gets called for both cases (CustomType and int).

What I'm basically trying to achieve is to have a custom `from_json` and `to_json` get called only when used with my `CustomType` and use the default serialization for `int`s.

**Comment by falbrechtskirchinger** [comment]
To the compiler, `CustomType` is still an `int` and will be treated as such. You could define a `struct` that implicitly converts to `int`.
